api/utils/keys/piv/yubikey.go

```diff
@@ -634,14 +561,90 @@ func (c *sharedPIVConnection) connect() (func(), error) {
 	return release, nil
 }
 
-func (c *sharedPIVConnection) privateKey(slot piv.Slot, public crypto.PublicKey, auth piv.KeyAuth) (crypto.PrivateKey, error) {
+type promptTouch func(ctx context.Context) error
+
+func (c *sharedPIVConnection) sign(ctx context.Context, ref *hardwarekey.PrivateKeyRef, auth piv.KeyAuth, promptTouch promptTouch, rand io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
+	pivSlot, err := parsePIVSlot(ref.SlotKey)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
 	release, err := c.connect()
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 	defer release()
-	privateKey, err := c.conn.PrivateKey(slot, public, auth)
-	return privateKey, trace.Wrap(err)
+
+	c.exclusiveOperationMu.RLock()
+	defer c.exclusiveOperationMu.RUnlock()
+
+	// Prepare the key and perform the signature with the same connection.
+	// Closing the connection in between breaks the underlying PIV handle.
+	priv, err := c.conn.PrivateKey(pivSlot, ref.PublicKey, auth)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	signer, ok := priv.(crypto.Signer)
+	if !ok {
+		return nil, trace.BadParameter("private key type %T does not implement crypto.Signer", priv)
+	}
+
+	return abandonableSign(ctx, signer, promptTouch, rand, digest, opts)
+}
+
+// abandonableSign extends [sharedPIVConnection.sign] to handle context, allowing the
+// caller to stop waiting for the result if the provided context is canceled.
+//
+// This is necessary for hardware key signatures which sometimes require touch from the
+// user to complete, which can block program termination.
+//
+// Important: this function only abandons the signer.Sign result, doesn't cancel it.
+func abandonableSign(ctx context.Context, signer crypto.Signer, promptTouch promptTouch, rand io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
+	ctx, cancel := context.WithCancelCause(ctx)
+	defer cancel(nil)
+
+	if promptTouch != nil {
+		go func() {
+			// There is no built in mechanism to prompt for touch on demand, so we simply prompt for touch after
+			// a short duration in hopes of lining up with the actual YubiKey touch prompt (flashing key). In the
+			// case where touch is cached, the delay prevents the prompt from firing when it isn't needed.
+			select {
+			case <-time.After(signTouchPromptDelay):
+				err := promptTouch(ctx)
+				if err != nil {
+					cancel(err)
+				}
+			case <-ctx.Done():
+				// prompt cached or signature canceled, skip prompt.
+			}
+		}()
+	}
+
+	type signResult struct {
+		signature []byte
+		err       error
+	}
+
+	signResultCh := make(chan signResult)
+	go func() {
+		defer close(signResultCh)
+		if err := ctx.Err(); err != nil {
+			return
+		}
+		signature, err := signer.Sign(rand, digest, opts)
+		select {
+		case <-ctx.Done():
+		case signResultCh <- signResult{signature: signature, err: trace.Wrap(err)}:
+		}
+	}()
+
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case result := <-signResultCh:
```

Comment:
Is it possible for this to race with `defer close(signResultCh)` when ctx is canceled?
Even if `signResultCh`, we should check the context to ensure go didn't pick the select case randomly 

> A `select` blocks until one of its cases can run, then it executes that case. It chooses one at random if multiple are ready.